### ISSUE TITLE
Azure SDK fails to decode non-ASCII text in API error messages

### ISSUE THREAD

**Issue by dharmab** [Azure/azure-sdk-for-python#9763]
- **Package Name**:  azure-cli
- **Package Version**: 2.0.77
- **Operating System**: Debian based Docker container
- **Python Version**: 3.8

**Describe the bug**

Azure SDK fails to decode some errors from the Azure API:

```
2020-01-30T17:53:23.6345580Z   File "/infrastructure/src/azutil/compute.py", line 233, in create_compute_object
2020-01-30T17:53:23.6345758Z     azutil.deployment.deploy_template(
2020-01-30T17:53:23.6345912Z   File "/infrastructure/src/azutil/deployment.py", line 53, in deploy_template
2020-01-30T17:53:23.6346241Z     if "OSProvisioningTimedOut" in str(e):
2020-01-30T17:53:23.6347151Z   File "/.venv/lib/python3.8/site-packages/msrestazure/azure_exceptions.py", line 193, in __str__
2020-01-30T17:53:23.6347501Z     return str(self.error)
2020-01-30T17:53:23.6348004Z   File "/.venv/lib/python3.8/site-packages/msrestazure/azure_exceptions.py", line 120, in __str__
2020-01-30T17:53:23.6349004Z     return error_bytes.decode('ascii')
2020-01-30T17:53:23.6349495Z UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 910: ordinal not in range(128)
```

**To Reproduce**
1. Hard to reproduce, since you have to get a pretty specific error.

**Expected behavior**

Azure SDK should correctly decode all Azure API errors and not throw UnicodeDecodeError.

**Comment by dw511214992** [comment]
We will follow this issue.

**Comment by jongio** [comment]
@dharmab - Just want to confirm.  You are seeing this with the CLI, not the SDK, correct?  Do you have a command that you run to consistently repro?

**Comment by dharmab** [comment]
No, this is a stack trace from one of our Python scripts that calls the
SDK, not from the az CLI. Unfortunately we've only seen it happen once and
can't consistently reproduce it, because it also requires Azure's server
side to throw an error with specific content.
--
Dharma Bellamkonda

**Comment by dharmab** [comment]
I think the error is in this file, actually: https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_exceptions.py

It looks like it's trying to decode strings as ascii. Which will fail if any character other than the english alphabet is present in an error message.

**Comment by mayurid** [comment]
@dharmab : Thanks for the report. Can you let us know what Python SDKs are you using?

**Comment by dharmab** [comment]
Here's the imports I'm using, does that help?

```
from azure.mgmt.resource import ResourceManagementClient
from azure.mgmt.resource.resources.models import DeploymentProperties
from msrestazure.azure_exceptions import CloudError
```

and the code I use:

```
def deploy_template(
    deployment_name: str,
    *,
    resource_group_name: str,
    arm_template: dict,
    resource_client: ResourceManagementClient,
) -> None:
    """
    Deploy an Azure Resource Manager template into the given resource group.

    :param deployment_name: A short name for the deployment. A timestamp will be suffixed to this name.
    :param resource_group_name: The resource group to deploy into
    :param arm_template: The ARM template to deploy
    :param resource_client: Azure resource management client
    """
    deployment_name = f"{deployment_name}-" + datetime.datetime.now().strftime(
        "%Y-%m-%d-%H%M"
    )

    deployment_arguments = {
        "resource_group_name": resource_group_name,
        "deployment_name": deployment_name,
        "properties": DeploymentProperties(
            template=arm_template, parameters={}, mode="Incremental"
        ),
        "polling": True,
    }
    try:
        logger.info('Validating template for deployment "{}"'.format(deployment_name))
        resource_client.deployments.validate(**deployment_arguments)
        logger.info('Launching deployment "{}"'.format(deployment_name))
        deployment_poller = resource_client.deployments.create_or_update(
            **deployment_arguments
        )
        azutil.wait.wait_for_long_azure_operation(
            f'Deployment "{deployment_name}"', poller=deployment_poller
        )
    except CloudError as e:
        if "OSProvisioningTimedOut" in str(e):
            logger.warning(
                "Deployment failed due to OS provisioning timeout. "
                "This can be caused by a fatal VM provisioning error, but can also occur when all VMs are functioning correctly. "
                "Attempting to continue..."
            )
        else:
            logger.exception("A problem occurred while deploying the ARM template")
            raise e
```

**Comment by alexagr** [comment]
I experience the same issue - I'm using a script written in Python 3 (and Python Azure SDK), to manage my Azure VMs. When I'm trying to create more VMs that my subscription limits allow, Python Azure SDK throws the following exception:
```
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 512
```

I was able to solve the problem by applying the following patch:
```
--- azure_exceptions.py.old 2020-02-20 15:55:55.293712427 +0000
+++ azure_exceptions.py 2020-02-20 15:54:17.437378094 +0000
@@ -117,7 +117,15 @@
             for error_info in self.additionalInfo:
                 error_str += "\n\t{}".format(str(error_info).replace("\n", "\n\t"))
         error_bytes = error_str.encode()
-        return error_bytes.decode('ascii')
+        try:
+            return error_bytes.decode('ascii')
+        except Exception:
+            pass
+        try:
+            return error_bytes.decode('utf-8')
+        except Exception:
+            pass
+        return error_str 
```

I'm using the latest version of Python Azure libraries installed via "pip3 install --upgrade azure":
```
ubuntu@alex-ubuntu-1:~$ pip3 freeze | grep azure
You are using pip version 8.1.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
azure==4.0.0
azure-applicationinsights==0.1.0
azure-batch==4.1.3
azure-common==1.1.24
azure-cosmosdb-nspkg==2.0.2
azure-cosmosdb-table==1.0.6
azure-datalake-store==0.0.48
azure-eventgrid==1.3.0
azure-graphrbac==0.40.0
azure-keyvault==1.1.0
azure-loganalytics==0.1.0
azure-mgmt==4.0.0
azure-mgmt-advisor==1.0.1
azure-mgmt-applicationinsights==0.1.1
azure-mgmt-authorization==0.50.0
azure-mgmt-batch==5.0.1
azure-mgmt-batchai==2.0.0
azure-mgmt-billing==0.2.0
azure-mgmt-cdn==3.1.0
azure-mgmt-cognitiveservices==3.0.0
azure-mgmt-commerce==1.0.1
azure-mgmt-compute==4.6.2
azure-mgmt-consumption==2.0.0
azure-mgmt-containerinstance==1.5.0
azure-mgmt-containerregistry==2.8.0
azure-mgmt-containerservice==4.4.0
azure-mgmt-cosmosdb==0.4.1
azure-mgmt-datafactory==0.6.0
azure-mgmt-datalake-analytics==0.6.0
azure-mgmt-datalake-nspkg==3.0.1
azure-mgmt-datalake-store==0.5.0
azure-mgmt-datamigration==1.0.0
azure-mgmt-devspaces==0.1.0
azure-mgmt-devtestlabs==2.2.0
azure-mgmt-dns==2.1.0
azure-mgmt-eventgrid==1.0.0
azure-mgmt-eventhub==2.6.0
azure-mgmt-hanaonazure==0.1.1
azure-mgmt-iotcentral==0.1.0
azure-mgmt-iothub==0.5.0
azure-mgmt-iothubprovisioningservices==0.2.0
azure-mgmt-keyvault==1.1.0
azure-mgmt-loganalytics==0.2.0
azure-mgmt-logic==3.0.0
azure-mgmt-machinelearningcompute==0.4.1
azure-mgmt-managementgroups==0.1.0
azure-mgmt-managementpartner==0.1.1
azure-mgmt-maps==0.1.0
azure-mgmt-marketplaceordering==0.1.0
azure-mgmt-media==1.0.0
azure-mgmt-monitor==0.5.2
azure-mgmt-msi==0.2.0
azure-mgmt-network==2.7.0
azure-mgmt-notificationhubs==2.1.0
azure-mgmt-nspkg==3.0.2
azure-mgmt-policyinsights==0.1.0
azure-mgmt-powerbiembedded==2.0.0
azure-mgmt-rdbms==1.9.0
azure-mgmt-recoveryservices==0.3.0
azure-mgmt-recoveryservicesbackup==0.3.0
azure-mgmt-redis==5.0.0
azure-mgmt-relay==0.1.0
azure-mgmt-reservations==0.2.1
azure-mgmt-resource==2.2.0
azure-mgmt-scheduler==2.0.0
azure-mgmt-search==2.1.0
azure-mgmt-servicebus==0.5.3
azure-mgmt-servicefabric==0.2.0
azure-mgmt-signalr==0.1.1
azure-mgmt-sql==0.9.1
azure-mgmt-storage==2.0.0
azure-mgmt-subscription==0.2.0
azure-mgmt-trafficmanager==0.50.0
azure-mgmt-web==0.35.0
azure-nspkg==3.0.2
azure-servicebus==0.21.1
azure-servicefabric==6.3.0.0
azure-servicemanagement-legacy==0.20.6
azure-storage-blob==1.5.0
azure-storage-common==1.4.2
azure-storage-file==1.4.0
azure-storage-queue==1.4.0
msrestazure==0.6.2
```

**Comment by dharmab** [comment]
Nice patch @alexagr. Some quick suggestions;

- [Unicode is backwards compatible with ASCII](https://en.wikipedia.org/wiki/UTF-8#Description), so I believe that the `decode('ascii')` call can be entirely removed.
- [You catch catch UnicodeDecodeException instead of Exception](https://docs.python.org/3/library/exceptions.html#UnicodeDecodeError). Since this is a fairly deep piece of library code, a specific catch rather than a general one is probably a safer choice.

If you open a PR into msrestazure I'd be happy to review (although I don't work for MS, so it'd just be some peer feedback).

By the way, the offending byte in your error turned out to be the `â` character- i.e. non-English text.

**Comment by lmazuel** [comment]
Fixed in msrestazure 0.6.3
https://github.com/Azure/msrestazure-for-python#2020-03-17-version-063
